fix saving and loading model including weights

spirousschuh opened the pull request:
## Motivation and Context

I encountered that saving a model including its weights and loading it again did not work for me the way that I would expect it. Therefore, I wrote a unit test that fails in the current version of the repo and I propose a code fix. 

**Detailed Problem Description**
From my perspective the underlying problem is that when saving a model using the function `tabpfn.model.loading.save_tabpfn_model` saves a `tabpfn.model.config.ModelConfig` together with the torch model. However, the model loading routine uses the `ModelConfig.upgrade_config` method that just works with dictionaries.

**The fix**
During the loading routine I decided to convert a `ModelConfig` to a `dict` in the `ModelConfig.upgrade_config` function. But of course one could fix the saving routine to save a dict instead of a `ModelConfig`.

---

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?
Yes. I provide a unit test for the regressor interface.
---

## Checklist

-   [x] The changes have been tested locally.
-   [x] Documentation has been updated (if the public API or usage changes).
-   [x] A entry has been added to `CHANGELOG.md` (if relevant for users).
-   [x] The code follows the project's style guidelines.
-   [x] I have considered the impact of these changes on the public API.

---

spirousschuh commented on tests/test_regressor_interface.py:
```diff
@@ -585,3 +589,40 @@ def test_initialize_model_variables_regressor_sets_required_attributes() -> None
+        save_tabpfn_model(regressor, save_path)
+
+        # Load the model state
+        with torch.serialization.safe_globals([ModelConfig]):
+            model, config, criterion = initialize_tabpfn_model(
+                save_path, "regressor", fit_mode="low_memory"
+            )
```
As the method is wrapped in the project's own loading / initializing function one might move the context manager inside of it. But unfortunately one needs it.

noahho commented:
@spirousschuh thanks for fixing this! I just merged with main which was quite a change, would you like to double-check?

noahho commented:
Sorry I have by now deviated a lot from your initial PR. Is it right that test_load_fitted_model is testing your case already?

noahho commented:
@spirousschuh okay great. Maybe the test file for saving and loading the model is the most appropriate place for the test - would be great to add that test there?

spirousschuh commented:
@noahho I added the test as you suggested and I added a small twist to the PR. Now, the `save_tabpfn_model` internally converts the `ModelConfig` to a `dict` before saving. This avoids adding the context manager to the loading that allows for loading custom classes. Now, we would not need the feature of the loading function to allow for loading `ModelConfig` anymore. Do you want me to remove it?

noahho commented:
Thanks a lot, also for adding that test. The fix now seems simpler, so removing the previous fix seems like a good idea. 
Are you familiar with how to ruff the code properly, otherwise I can give it a go?
Also, was there a typing issue for the input to the save / load function input?

noahho reviewed (approved):
Amazing, thanks a lot for this!

noahho commented:
/gemini review
